### PR TITLE
Adding Network Function subdirectory

### PR DESCRIPTION
**Pull request by paulfhncw**
First cut of network function dummy code
Signed-off-by: Paul Hough <paul.f.hough@intel.com>

### CODE REVIEW

File: network_function/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+Network Function Skeleton Code
+
+This code attachs to an IPU network interface via a raw socket.
```
Comment (Eoghan1232) [inline network_function/README.md:3]:
```suggestion
This code attaches to an IPU network interface via a raw socket.
```

File: network_function/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+To compile 
+gcc -o nf nf.c
+
+to run
+./nf
+
+To set up the test environment
+
+Add a new nsamespace
```
Comment (Eoghan1232) [inline network_function/README.md:17]:
```suggestion
Add a new namespace
```

File: network_function/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ip netns exec ns1 ip link set tap1 up
+ip netns exec ns2 ip link set tap2 up
+
+Run code:
+ip netns exec nfns ./nf br-tap1 br-tap2
+
+Run test:
+ip netns exec ns1 ping -c1 10.1.1.5
+
+Exspected Behavior
```
Comment (Eoghan1232) [inline network_function/README.md:49]:
```suggestion
Expected Behavior
```

File: network_function/README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+packet could will increment, packet output to screen, source MAC printed.
+packet relayed out second interface and visible via tcpdump command
+sudo ip netns exec ns2 tcpdump -vi tap2
```
Comment (Eoghan1232) [inline network_function/README.md:52]:
```suggestion
packet will increment, packet output to screen, source MAC printed.
packet relayed out second interface and visible via tcpdump command
sudo ip netns exec ns2 tcpdump -vi tap2
```

**Review by Eoghan1232 — APPROVED** [review]
some small nits - otherwise lgtm

File: network_function/nf.c [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        // Print the source MAC address
+        printf("Source MAC Address: %02x:%02x:%02x:%02x:%02x:%02x\n",
+           sourceMac[0], sourceMac[1], sourceMac[2], sourceMac[3], sourceMac[4], sourceMac[5]);
+
+
+        if (length == -1) {
+            perror("Failed to receive packet");
+            close(rawSocket);
+            exit(1);
+        }
```
Comment (palade) [inline network_function/nf.c:96]: This could be moved after `receivefrom` if that function failed

File: network_function/nf.c [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    sa_out.sll_ifindex = ifr_out.ifr_ifindex;
+    static int i = 0;
+
+    // Receive and send packets
+    while (1) {
+        char buffer[65536];
+        ssize_t length;
+
+        // Receive packet
+        length = recvfrom(rawSocket, buffer, sizeof(buffer), 0, NULL, NULL);
```
Comment (palade) [inline network_function/nf.c:74]: An assumption is made here that the packet size is 65536 - if it is larger then this will be ignored.

**Review by palade — COMMENTED** [review]
Few small comments

**Review by davecremins — CHANGES_REQUESTED** [review]
Needs a container/pod manifest for deployment into a K8s cluster

File: network_function/README.md (hunk 1 above)
Comment (paulfhncw) [reply network_function/README.md:3]: updated

Comment (paulfhncw) [reply network_function/README.md:3]: updated

File: network_function/README.md (hunk 2 above)
Comment (paulfhncw) [reply network_function/README.md:17]: updated

File: network_function/README.md (hunk 3 above)
Comment (paulfhncw) [reply network_function/README.md:49]: updated

File: network_function/README.md (hunk 4 above)
Comment (paulfhncw) [reply network_function/README.md:52]: updated

File: network_function/nf.c [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        printf("%02x ", (unsigned char)packet[i]);
+        if ((i + 1) % 16 == 0) {
+            printf("\n");
+        }
+    }
+    printf("\n");
+}
+
+int main() {
+    int rawSocket;
```
[the remark this answers is not in the document]
Comment (paulfhncw) [reply network_function/nf.c:28]: raw socket is type of socket which transfers both the ip header and payload to user space.

File: network_function/nf.c (hunk 5 above)
Comment (paulfhncw) [reply network_function/nf.c:96]: yeah, should be moved. Code updated.

File: network_function/nf.c (hunk 6 above)
Comment (paulfhncw) [reply network_function/nf.c:74]: yes, agreed, as it is sample code - i think this is ok. But i will add a comment to the Readme.

File: network_function/README.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
+Network Function Skeleton Code
```
Comment (ahalimx86) [inline network_function/README.md:1]: This README.md is missing markdown formatting

File: network_function/dockerfile [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Specify the base image
+FROM gcc:latest
+# Copy the program files
+COPY nf.c /app/
+# Set the working directory
+WORKDIR /app
+# Compile the program
+RUN gcc -o nf nf.c
+# Set the entry point
+RUN chmod +x nf
```
Comment (ahalimx86) [inline network_function/dockerfile:10]: You can append this command with && into previous single RUN directive to reduce one additional layer in the container image.

File: network_function/dockerfile [hunk 10]
```diff
@@ -0,0 +1,11 @@
+# Specify the base image
```
Comment (ahalimx86) [inline network_function/dockerfile:1]: Missing copyright and licensing heading

File: network_function/nf.c [hunk 11]
```diff
@@ -0,0 +1,108 @@
+#include <stdio.h>
```
Comment (ahalimx86) [inline network_function/nf.c:1]: Missing copyright and licensing heading

Comment (paulfhncw) [reply network_function/nf.c:1]: added

File: network_function/dockerfile (hunk 10 above)
Comment (paulfhncw) [reply network_function/dockerfile:1]: added

File: network_function/dockerfile (hunk 9 above)
Comment (paulfhncw) [reply network_function/dockerfile:10]: changed

**Review by davecremins — APPROVED** [review]

File: network_function/README.md [hunk 12]
```diff
@@ -7,7 +7,7 @@ This provides a "transparent pipe" which can be extended to
 provide true "bump in the wire" network functions.
 This code supports packets up to 64K.
 
-To compile 
+**To compile**
```
Comment (Eoghan1232) [inline network_function/README.md:10]: use code blocks for what you expect to be run/executed

File: network_function/README.md [hunk 13]
```diff
@@ -7,7 +7,7 @@ This provides a "transparent pipe" which can be extended to
...
 gcc -o nf nf.c
```
Comment (Eoghan1232) [inline network_function/README.md:11]: same as above.

File: network_function/README.md (hunk 12 above)
Comment (paulfhncw) [reply network_function/README.md:10]: updated

File: network_function/README.md (hunk 13 above)
Comment (paulfhncw) [reply network_function/README.md:12]: updated

**Review by Eoghan1232 — APPROVED** [review]

**Review by palade — APPROVED** [review]

**Pull request merged**